kv/kvserver: TestLeasePreferencesDuringOutage failed

kv/kvserver.TestLeasePreferencesDuringOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19098817?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19098817?buildTab=artifacts#/) on master @ [a3c6f5363dd11bbcc938543c5e7e7b1b0f2c589f](https://github.com/cockroachdb/cockroach/commits/a3c6f5363dd11bbcc938543c5e7e7b1b0f2c589f):


```
=== RUN   TestLeasePreferencesDuringOutage
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/819b665086d80deae71a56aaf1b41473/logTestLeasePreferencesDuringOutage2171364851
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1235: 
        	Error Trace:	pkg/kv/kvserver/client_lease_test.go:1235
        	Error:      	Expected nil, but got: &kvpb.Error{UnexposedTxn:(*roachpb.Transaction)(nil), OriginNode:0, EncodedError:errorspb.EncodedError{Error:(*errorspb.EncodedError_Wrapper)(0x400b7a84e8)}, Index:(*kvpb.ErrPosition)(nil), Now:hlc.ClockTimestamp{WallTime:1741914245174897387, Logical:0}}
        	Test:       	TestLeasePreferencesDuringOutage
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/819b665086d80deae71a56aaf1b41473/logTestLeasePreferencesDuringOutage2171364851
--- FAIL: TestLeasePreferencesDuringOutage (2.36s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeasePreferencesDuringOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48559

kv/kvserver.TestLeasePreferencesDuringOutage [failed](https://tanzanite.cluster.engflow.com/invocations/default/57c971ff-77bf-4a19-874d-9e6767b26b74?testReportRun=7&testReportShard=26&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [b53856cc8cebf5fe28a8fca3940f5603b5b2d833](https://github.com/cockroachdb/cockroach/commits/b53856cc8cebf5fe28a8fca3940f5603b5b2d833):


```
=== RUN   TestLeasePreferencesDuringOutage
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestLeasePreferencesDuringOutage433707978
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1237: condition failed to evaluate within 45s: from client_lease_test.go:1245: expected no leaseholder in region=us, but found {repl=(n1,s1):1 seq=3 start=1742104355.204891623,1 epo=1 min-exp=0,0 pro=1742104649.220274066,0 acq=Request 1742104695.322579425,0 1742104695.322579425,0 VALID  liveness(nid:1 epo:1 exp:1742104699.215637696,0) 1742104355.204891623,1 {0 0,0}}
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestLeasePreferencesDuringOutage433707978
--- FAIL: TestLeasePreferencesDuringOutage (48.39s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=26</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeasePreferencesDuringOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestLeasePreferencesDuringOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19116324?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19116324?buildTab=artifacts#/) on master @ [21eeb1b2ff2d08ed061a2c70083af9b9656dba15](https://github.com/cockroachdb/cockroach/commits/21eeb1b2ff2d08ed061a2c70083af9b9656dba15):


```
=== RUN   TestLeasePreferencesDuringOutage
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/819b665086d80deae71a56aaf1b41473/logTestLeasePreferencesDuringOutage2760156519
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1235: 
        	Error Trace:	pkg/kv/kvserver/client_lease_test.go:1235
        	Error:      	Expected nil, but got: &kvpb.Error{UnexposedTxn:(*roachpb.Transaction)(nil), OriginNode:0, EncodedError:errorspb.EncodedError{Error:(*errorspb.EncodedError_Wrapper)(0x40115c3cc8)}, Index:(*kvpb.ErrPosition)(nil), Now:hlc.ClockTimestamp{WallTime:1742196165075187007, Logical:0}}
        	Test:       	TestLeasePreferencesDuringOutage
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/819b665086d80deae71a56aaf1b41473/logTestLeasePreferencesDuringOutage2760156519
--- FAIL: TestLeasePreferencesDuringOutage (3.27s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeasePreferencesDuringOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

```
--- FAIL: TestLeasePreferencesDuringOutage (2.36s)
    test_log_scope.go:165: test logs captured to: /tmp/cockroach/_tmp/b26415f860e738404d72d15f6087790e/logTestLeasePreferencesDuringOutage325213443
    test_log_scope.go:76: use -show-logs to present logs inline
    client_lease_test.go:1235:
                Error Trace:    pkg/kv/kvserver/client_lease_test.go:1235
                Error:          Received unexpected error:
                                replica unavailable: (n0,s0):? unable to serve request to r75:{-} [<no replicas>, next=0, gen=0]: lost quorum (down: ); closed timestamp: 0,0 (1970-01-01 00:00:00); raft status: <nil>: replica has been leaderless for 1m0s
                                (1) replica unavailable: (n0,s0):? unable to serve request to r75:{-} [<no replicas>, next=0, gen=0]: lost quorum (down: ); closed timestamp: 0,0 (1970-01-01 00:00:00); raft status: <nil>: replica has been leaderless for 1m0s
                                Wraps: (2)
                                  | (opaque error wrapper)
                                  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
                                  | reportable 0:
                                  |
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.replicaUnavailableError
                                  |     pkg/kv/kvserver/replica_circuit_breaker.go:268
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableError
                                  |     pkg/kv/kvserver/replica_circuit_breaker.go:277
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.NewLeaderlessWatcher
                                  |     pkg/kv/kvserver/replica.go:214
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.newUninitializedReplicaWithoutRaftGroup
                                  |     pkg/kv/kvserver/replica_init.go:248
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.newUninitializedReplica
                                  |     pkg/kv/kvserver/replica_init.go:105
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).tryGetOrCreateReplica
                                  |     pkg/kv/kvserver/store_create_replica.go:215
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).getOrCreateReplica
                                  |     pkg/kv/kvserver/store_create_replica.go:74
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).acquireSplitLock
                                  |     pkg/kv/kvserver/replica_raft.go:2969
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).maybeAcquireSplitMergeLock
                                  |     pkg/kv/kvserver/replica_raft.go:2958
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).runPostAddTriggersReplicaOnly
                                  |     pkg/kv/kvserver/replica_app_batch.go:237
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).Stage
                                  |     pkg/kv/kvserver/replica_app_batch.go:157
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.mapCmdIter
```

So it looks like the `LeaderlessWatcher` is going at it a little too hard. I'll bisect. The commit that introduces this watcher doesn't seem to flake, at least not nearly as readily.

--- a/pkg/kv/kvserver/client_lease_test.go
+++ b/pkg/kv/kvserver/client_lease_test.go
@@ -1112,7 +1112,11 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 	// clock jumps. Disable the suspect timer to prevent them becoming suspect
 	// when we bump the clocks.
 	liveness.TimeAfterNodeSuspect.Override(ctx, sv, 0)
+
 	timeUntilNodeDead := liveness.TimeUntilNodeDead.Get(sv)
+	// We'll jump the clock forward and don't accidentally want the watcher to
+	// fire.
+	kvserver.ReplicaLeaderlessUnavailableThreshold.Override(ctx, sv, 5*timeUntilNodeDead)
 
 	for i := 0; i < numNodes; i++ {
 		serverArgs[i] = base.TestServerArgs{
@@ -1232,7 +1236,7 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 	ba := &kvpb.BatchRequest{}
 	ba.Add(getArgs(key))
 	_, pErr := tc.Servers[0].DistSenderI().(kv.Sender).Send(ctx, ba)
-	require.Nil(t, pErr)
+	require.NoError(t, pErr.GoError())
 
 	testutils.SucceedsSoon(t, func() error {
 		// Validate that the lease transferred to a preferred locality. n4 (us) and

